File: Data Structures & Algorithms/missing-number/submission-4.py
class Solution:
    def missingNumber(self, nums: List[int]) -> int:
        '''
        xor approach
        xor = len of nums
        loop over the array
        xor i with xor
        xor arr[i] with xor
        '''

        xor = len(nums)

        for i in range(len(nums)):
            xor ^= i ^ nums[i]

        return xor

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-4.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] output = new int[n];
        int prod = 1;

        for (int i = 0; i < n; i++) {
            output[i] = prod;    
            prod *= nums[i];
        }
        
        prod = 1;

        for (int j = n - 1; j >= 0; j--) {
            output[j] *= prod;
            prod *= nums[j];
        }

        return output;
    }
}  
